nixos/modules/virtualisation/google-compute-image.nix

```diff
@@ -132,40 +132,48 @@ in
       after = [ "network-online.target" "ip-up.target" ];
       wants = [ "network-online.target" "ip-up.target" ];
 
-      script = let wget = "${pkgs.wget}/bin/wget --retry-connrefused -t 15 --waitretry=10 --header='Metadata-Flavor: Google'"; in
+      script = let wget = "${pkgs.wget}/bin/wget --retry-connrefused -t 15 --waitretry=10 --header='Metadata-Flavor: Google'";
+                   mktemp = "mktemp --tmpdir=/run"; in
         ''
           # When dealing with cryptographic keys, we want to keep things private.
           umask 077
           # Don't download the SSH key if it has already been downloaded
           if ! [ -e /root/.ssh/authorized_keys ]; then
-                echo "obtaining SSH key..."
-                mkdir -m 0700 -p /root/.ssh
-                ${wget} -O /root/authorized-keys-metadata http://metadata.google.internal/0.1/meta-data/authorized-keys
-                if [ $? -eq 0 -a -e /root/authorized-keys-metadata ]; then
-                    cat /root/authorized-keys-metadata | cut -d: -f2- > /root/key.pub
-                    if ! grep -q -f /root/key.pub /root/.ssh/authorized_keys; then
-                        cat /root/key.pub >> /root/.ssh/authorized_keys
-                        echo "new key added to authorized_keys"
-                    fi
-                    chmod 600 /root/.ssh/authorized_keys
+              echo "obtaining SSH key..."
+              mkdir -m 0700 -p /root/.ssh
+              AUTH_KEYS=$(${mktemp}) && {
```

Comment:
Yes, unless there is actually something you can do about an error, just let the whole thing fail as early as possible. To me it seems that what you really want is to verify that whatever wget downloaded looks like a key file (i.e., not garbage or something else). Testing for non-zero files doesn't seem too useful to me.

On an unrelated note, I don't quite see how writing tempfiles directly under `/run` is any better than `/tmp`, unless you use a dedicated run directory for the service (i.e., `/run/google-image`). Surely, if they can race you on `/tmp`, they can race on `/run`, too (or am I missing something?).

For private keys, perhaps it might be worth considering writing to `/run/keys`, even, as that is sure to never touch the disk (ramfs). For pubkeys, the most important thing seems to me to avoid someone replacing the key with their own, not keeping them from seeing your key (what would they do with it, give you access to their box?). The fact that the pubkey is acquired over plain http would worry me, but I'm unsure how much of a concern that really is here.

Anyway, just my 2 NOK.